src/Widgets/AbstractAppContainer.vala

```diff
@@ -163,88 +157,57 @@ namespace AppCenter {
-                        /* Uninstall button will show */
-                        action_button.no_show_all = true;
-                        action_button.hide ();
-
-                        if (!is_os_updates) {
-                            uninstall_button.no_show_all = false;
-                            uninstall_button.show_all ();
-                        }
-                    } else {
-                        /* No Uninstall action in list view */
-                        action_stack.no_show_all = true;
-                        action_stack.hide ();
-                    }
-
-                    if (show_open && package.get_can_launch ()) {
-                        open_button.no_show_all = false;
-                        open_button.show ();
-                    } else {
-                        open_button.no_show_all = true;
-                        open_button.hide ();
-                    }
+                    set_button_state (uninstall_button, show_uninstall && !is_os_updates);
+                    set_button_state (action_button, false);
+                    set_button_state (open_button, show_open && package.get_can_launch ());
 
                     break;
-
                 case AppCenterCore.Package.State.UPDATE_AVAILABLE:
-                    action_button.can_purchase = false;
                     action_button.label = _("Update");
 
-                    action_button.no_show_all = false;
-                    action_button.show_all ();
-
-                    if (show_open && package.get_can_launch ()) {
-                        open_button.no_show_all = false;
-                        open_button.show ();
-                    } else {
-                        open_button.no_show_all = true;
-                        open_button.hide ();
-                    }
+                    set_button_state (uninstall_button, show_uninstall && !is_os_updates);
+                    set_button_state (action_button, true);
+                    set_button_state (open_button, false);
 
                     break;
-
                 case AppCenterCore.Package.State.INSTALLING:
                 case AppCenterCore.Package.State.UPDATING:
                 case AppCenterCore.Package.State.REMOVING:
-                    progress_bar.no_show_all = false;
-                    progress_bar.show ();
-                    action_stack.set_visible_child_name ("progress");
+                    set_button_state (uninstall_button, false);
+                    set_button_state (action_button, false);
+                    set_button_state (open_button, false);
 
-                    open_button.no_show_all = true;
-                    open_button.hide ();
+                    action_stack.set_visible_child_name ("progress");
                     break;
 
                 default:
                     assert_not_reached ();
             }
         }
 
+        private void set_button_state (Gtk.Widget widget, bool show) {
```

Comment:
Since this function isn't using anything from the instance, it could perhaps be `static`?

I guess, you could also make use of `visible` instead of `show ()` and `hide ()`:
widget.visible = show;

But it's really optional, you can keep how it is, though I would like to change `show` to `show_all`